File: backend/app/core/upgrades/abilities/double_dash.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class DoubleDash(UpgradeImplementation):
    id = 'double_dash'
    name = 'Double Dash'
    description = 'Store 2 dashes.'
    rarity = 'epic'
    type = 'ability'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        appliers.effects.register_effect('double_dash', {})

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/abilities/double_dash.ts
import { Upgrade, type UpgradeDef, type UpgradeContext } from '../Upgrade'
import { RarityID, UpgradeTypeID } from '../../data/ID'

export const DoubleDashDef: UpgradeDef = {
  id: "double_dash",
  name: "Double Dash",
  description: "Store 2 dashes.",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.Ability,
  cost: 20,
  effect: "double_dash",
  stackable: false,
  dependentOn: [{ ids: ["dash_ability"] }],
}

export class DoubleDash extends Upgrade {
  onApply(ctx: UpgradeContext): void {
    super.onApply(ctx)
    ctx.player?.setMaxDashCharges(2)
  }
}
